cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/inference/Inference.kt

```diff
@@ -528,6 +539,52 @@ class Inference internal constructor(val start: Node, override val ctx: Translat
         this.scopeManager = ctx.scopeManager
         this.typeManager = ctx.typeManager
     }
+
+    fun inferReturnType(call: CallExpression?): Type {
+        if (call == null) {
+            return unknownType()
+        }
+
+        // Try to find out, if the supplied hint is part of an assignment.  If yes, we can use their
+        // type as the return type of the function
+        var targetType =
+            ctx.currentComponent.assignments.firstOrNull { it.value == call }?.target?.type
+        if (targetType != null && targetType !is UnknownType) {
+            return targetType
+        }
+
+        // Look for an "argument holder". These can be different kind of nodes
+        val holder =
+            ctx.currentComponent.allChildren<ArgumentHolder> { it.hasArgument(call) }.singleOrNull()
+        when (holder) {
+            is UnaryOperator -> {
+                // If it's a boolean operator, the return type is probably a boolean
+                if (holder.operatorCode == "!") {
+                    return call.language?.builtInTypes?.values?.firstOrNull { it is BooleanType }
+                        ?: unknownType()
+                }
+                // If it's a numeric operator, return the fist numeric type that we have
+                if (holder.operatorCode in listOf("+", "-", "++", "--")) {
```

Comment:
This is extremely language-specific.

At least for `+`, it could also be a string in a good amount of languages...

For languages with operator overloading, it could be anything.

Reply:
If the idea is to implement an inference heuristic that "maaaay" still be ok here, hmm do we have a better alternative?

Reply:
> This is extremely language-specific.
> 
> At least for `+`, it could also be a string in a good amount of languages...
> 
> For languages with operator overloading, it could be anything.

Note that this is a *unary* operator with a numeric operation, so the assumption that it is any numeric type is probably ok. We could extend this heuristic of course to first look for integer types, with the highest bidwidth, to over-approximate and then if none are found, use a float type. Not sure if this is really better though.

As for operator overloading, yes of course, there will be cases where we are completely wrong, but on the other hand, there is no possibility at all to deduce the correct type. Return types are not taken into account when resolving, so the "worst" thing we can do is to propagate a wrong type into the `assignedTypes` of the data flows that follow and that is ok-ish.